### DISCUSSION TITLE
Identity (autoincrement) key and unit of work commit

### DISCUSSION THREAD

**Discussion by martinobordin** [kgrzybek/modular-monolith-with-ddd#220] [Q&A] [answered]
Hi,
I see that the command i.e. [RegisterNewUserCommandHandler ](https://github.com/kgrzybek/modular-monolith-with-ddd/blob/90ab9b20a1c6e11700e8b357929c5f792b2b32bf/src/Modules/UserAccess/Application/UserRegistrations/RegisterNewUser/RegisterNewUserCommandHandler.cs)  is returning the Id of the new entity before the UnitOfWork decorator actually commits the changes.

This is ok if the Id is generated by the application (as in your case), but how to handle it in the case it's generated database side (identity keys).

Probably the command should call the unit of work commit itself.

Any thoughts?

**Comment by AndreiGanichev** [comment, marked as answer]
Hi, @martinobordin.
First of all I would like to emphasise that following concepts are essentially independent:
1. identity of Entity(DDD tactical pattern)
2. database primary key

> Some ORM tools want to deal with object identity on their own terms. To cure this, we need to use two identities. One of the identities is designed for the domain model and adheres to the requirements of the domain. The other is for Hibernate and is known as a surrogate identity...Remember, in this case it[surrogate identity] has nothing to do with the domain-specific identity. It is being created only for the sake of the ORM. 
(Vaughn Vernon. IDDD. Chapter 5. Entities. Surrogate key)

You can read more in [Entity Identity vs Database Primary Key](https://enterprisecraftsmanship.com/posts/entity-identity-vs-database-primary-key/) by Vladimir Khorikov.

But of corse these concepts have two important requirements in common: they have to be immutable and globally unique. So it's often when entity identity is used as primary key. And Vaughn Vernon writes:

> For Aggregate Root identity generation, I like to use its Repository. This seems like a natural location for identity generation.
(Vaughn Vernon. IDDD. Chapter 5. Entities.)

Returning to your question I think it's good separation of concerns to use decorator to manage database transaction. So to be able to return entity Id from a command handler Id generation must be a part of entities construction. So we can:
1. Add separate method generating Id to repository. Command handler generates Id in repository, passes Id to entity constructor, passes created entity to repository and returns Id. These approach is described in IDDD book Chapter 5. Entities.When the Timing of Identity Generation Matters.
2. Use [Hi/Lo algorithm](https://en.wikipedia.org/wiki/Hi/Lo_algorithm). There is [Hi/Lo algorithm in EF Core](https://docs.microsoft.com/en-us/dotnet/architecture/microservices/microservice-ddd-cqrs-patterns/infrastructure-persistence-layer-implementation-entity-framework-core#the-hilo-algorithm-in-ef-core) for example.
